### PR TITLE
gh-103091: Add PyUnstable_Type_AssignVersionTag

### PR DESCRIPTION
**Pull request by swtaarrs**
Add a new C-API function to eagerly assign a version tag to a PyTypeObject. Details in gh-103091.

### CODE REVIEW

File: Include/object.h [hunk 1]
```diff
@@ -861,6 +861,15 @@ static inline int PyType_CheckExact(PyObject *op) {
 #  define PyType_CheckExact(op) PyType_CheckExact(_PyObject_CAST(op))
 #endif
 
+#if !defined(Py_LIMITED_API)
```
Comment (encukou) [inline Include/object.h]: You can put the definition in `Include/cpython/object.h`, rather than use this `#ifdef` here. (See line 825.)

Comment (carljm) [reply Include/object.h]: See https://devguide.python.org/developer-workflow/c-api/index.html for details on this.

**Review by carljm — CHANGES_REQUESTED** [review]
Requesting changes for renaming to `PyUnstable_Type_AssignVersionTag` and moving to `Include/cpython/object.h`.

**Comment by Security2965** [comment]
It's fixed now so go & enjoy

**Comment by swtaarrs** [comment]
Thanks for the comments!

I have made the requested changes; please review again.

**Review by carljm — APPROVED** [review]
This looks good to me. @markshannon any concerns here?

**Review by carljm — CHANGES_REQUESTED** [review]
On second look here, I realized the updated name isn't quite what was requested; it should be `PyUnstable_Type_AssignVersionTag` (not just `PyUnstable_AssignVersionTag`).

**Comment by swtaarrs** [comment]
Sorry about that! I fixed the name and merged latest main.

I have made the requested changes; please review again.

**Review by carljm — APPROVED** [review]

**Comment by carljm** [comment]
Checked this in person with @markshannon and @ericsnowcurrently at the PyCon sprints; going ahead and merging.

**Pull request merged**